### DISCUSSION TITLE
How to build one Application for different branches and app bundles

### DISCUSSION THREAD

**Discussion by bobandrew** [codemagic-ci-cd/codemagic-docs#2152] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

I have one configured application in codemagic. And I would like to start build of this application for my different branches and bundles. In nutshell, I have one codebase, but for every my customer I create new branch with his branded application. So I wanted to use codemagic to build my different branches for different bundles and publish application updates to App Store and Google Play. Is this possible?

### Expected results

For example, I want to have API like:
POST /builds
appId = xxx
branch = xxx
bundle = xxx

In your documentation I didn't find 'bundle' parameter for /builds API call (https://docs.codemagic.io/rest-api/builds/)

Thanks in advance

### Actual results

no

**Comment by dtrdic** [comment]
Hi @bobandrew here's an example on how you would trigger multiple builds for multiple branches

```
          APP_ID="testappid1234324"  # Replace with your App ID
          BRANCHES=("main" "master" "dev")

          for BRANCH in ${BRANCHES[@]}; do
            echo "Triggering workflow $WORKFLOW for branch $BRANCH"

            curl -H "Content-Type: application/json" -H "x-auth-token: $CM_API_KEY" \
              --data '{
                "appId": "'${APP_ID}'",
                "workflowId": "android-workflow",
                "branch": "'${BRANCH}'"
              }' \
              https://api.codemagic.io/builds
          done
```
is the bundleID essential in your case if the workflowId(or a branch name) is different for every customer?

**Reply by bobandrew** [reply]
@dtrdic , yes, I found this call in your documentation, but I'm wondering how to pass 'bundle' in this call? I want to pass 'bundle' so that it will be published  in stores. Or I'm missing something and I don't  need to pass 'bundle'?

Thansk for you help!

**Reply by icarusdust** [reply, marked as answer]
Hi @bobandrew, sounds like you are white-labeling your app. In a nutshell, what you need is to store your client assets in a bucket, could be any of your preference (S3, GCP, Firebase etc.) then you will allow these assets to be downloaded during a build and by providing a client id, you will target a specific client assets to be downloaded. Then, they will replace the one that you have in your original repo by default. You will need to create environment variable groups that hold hold each client's specific env vars such as `BUNDLE_ID`:

```
curl -H "Content-Type: application/json" -H "x-auth-token: <your-auth-token>" \
      --data '{
          "appId": "<your-codemagic-app-id>", 
          "workflowId": "<your-codemagic-workflow-id>",
          "branch": "<branch-name>",
          "environment": { 
              "variables": { 
                  "CLIENT_ID": "<your-client-id>"
              },
              "groups": [
                  "<your-client-group-name>"
              ]
          }
        }' \
       https://api.codemagic.io/builds
```

So, when triggering a build on a specific client, their BUNDLE_ID env var will be shipped with this cURL request, and will replace the default bundle id inside `project.pbxproj` via `sed` stream editor utility:

```
- name: Set bundle id
  script: sed -i '' -e 's/PRODUCT_BUNDLE_IDENTIFIER \= [^\;]*\;/PRODUCT_BUNDLE_IDENTIFIER = '${BUNDLE_ID}';/' ios/Runner.xcodeproj/project.pbxproj
```

It is hard to explain all these steps here. I would suggest you to go through our documentation on how to white-label apps: https://docs.codemagic.io/yaml-quick-start/white-label-getting-started/

**Reply by bobandrew** [reply]
@icarusdust , great, I will try. Thanks a lot.

**Comment by dtrdic** [comment]
I'm tentatively closing the discussion due to a lack of activity. You can open the discussion again by replying to this message or opening a new discussion.
